### ISSUE TITLE
ENH: Fix the random state in `scipy.stats.multivariate_normal.cdf()`

### ISSUE THREAD

**Issue by Mv77** [scipy/scipy#16142]
### Is your feature request related to a problem? Please describe.

The CDF in the `scipy.stats.multivariate_normal` is computed using Alan Genz's algorithm. This is the main Fortran subroutine that is being used https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/stats/mvndst.f#L157

An issue is that the algorithm computes the relevant integrals through Monte Carlo simulations. I believe that scipy's implementation is not fixing the random state when calling the Fortran routine. This makes it so that multiple calls of `scipy.stats.multivariate_normal.cdf()` with exactly the same input can return different outputs. This is problematic for, e.g., maximum likelihood estimation of models that use the probabilities that come out of the method. 

### Describe the solution you'd like.

I'd like to be given the option to fix the seed of the Fortran simulations that are used to compute the CDF, so that I can get identical outputs when I call `scipy.stats.multivariate_normal.cdf()` with the same inputs.

R's package [`mvtnorm`](https://cran.r-project.org/web/packages/mvtnorm/index.html) appears to allow this.

**Comment by tupui** [comment]
Hi @Mv77 thank you for reporting. I removed the link to R since it's GPL which is not compatible with our license. Please don't mention anything about this package as we cannot use anything about it (not even read the code).

**Comment by rkern** [comment]
The PRNG is all in FORTRAN. It's entirely unrelated to our PRNGs. It maintains its [own state as `SAVE`d variables](https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/stats/mvndst.f#L1170). We can't really manipulate that state from the outside AFAIK.

We'd have to rewrite the FORTRAN code to call a C function that would call our `BitGenerator` that we'd set in the wrapper around the `mvnun_weighted()` call. Not especially easy to do with an f2py wrapper but certainly doable with a handwritten wrapper.

**Comment by rkern** [comment]
The author of `mvndst.f` has [newer methods implemented in MATLAB](http://www.math.wsu.edu/faculty/genz/software/software.html) (namely [QSIMVNV](http://www.math.wsu.edu/faculty/genz/software/matlab/qsimvnv.m)). They are QMC-based, so should be better quality. They still use a PRNG, but porting the MATLAB code will give us the opportunity to work with a `Generator` natively.

**Comment by ilayn** [comment]
I looked into this innocent looking Fortran code, and as you expected it is just terror. Fortunately it uses the randomized parts in only two places which are ok to be cleaned up. One of the use case is easily one of my new favorites. 

https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/stats/mvndst.f#L794

It is using a floating point number between (0, 1) multiplication to generate a random integer between `0` and `( NK + 1 - J )` then casting back to integer because Fortran. **Fantastic!**

I can free it from `SAVE` bonanza but I still don't quite understand how to get the Random C-API work in C. The extension examples are quite thin on that front. For example they all use the Python API to pick the PCG64 instance but I could not understand how to grab it from C. So if you have a working copy somewhere that does all in C and shows how to reach out to things like `shuffle` or `choice` that would make things much easier for me. Or if these are high-level interfaces, simple integer or double generation is also fine.

**Comment by ev-br** [comment]
https://github.com/ev-br/mc_lib/blob/master/mc_lib/rndm.pyx

Here's a worked example in cython , basically copy-pasted from numpy docs.

**Comment by ilayn** [comment]
Cython is fine but from C it turned out to be hairier than I expected.

**Comment by ev-br** [comment]
Yeah, C is always hairier :-). The main thing IIRC is the PyCapsule.get_pointer dance, and  this is not cython-specific?

**Comment by mdhaber** [comment]
See the plan suggested in https://github.com/scipy/scipy/pull/17410#issuecomment-2338786323.
